sdk/go/common/registry/resolve.go

```diff
@@ -0,0 +1,166 @@
+// limitations under the License.
+
+package registry
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/blang/semver"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+)
+
+// ResolvePackageFromName resolves a registry package from user input. User input may be in the following forms:
+//
+//	<source>/<publisher>/<name> -> [<source>/<publisher>/<name>]
+//
+//	<publisher>/<name>          -> [private/<publisher>/<name>, pulumi/<publisher>/<name>]
+//
+//	<name>                      -> [private/*/<name>, pulumi/*/<name>]
+//
+// The returned error will include [NotFoundError] if and only if ResolvePackageFromName
+// has determined that registry does not contain a matching package.
+//
+// If ResolvePackageFromName could not parse name into a fragment kind, it will return an
+// error that includes [InvalidIdentifierError].
+//
+// When ResolvePackageFromName is unable to find a corresponding package (and
+// [NotFoundError] is returned), a list of similar packages *may* be available from
+// calling [GetSuggestedPackages] on the returned error.
+func ResolvePackageFromName(
+	ctx context.Context, registry Registry, name string, version *semver.Version,
+) (apitype.PackageMetadata, error) {
+	parts := strings.Split(name, "/")
+	switch len(parts) {
+	case 3:
+		p, err := registry.GetPackage(ctx, parts[0], parts[1], parts[2], version)
+		return p, err
+	case 2:
+		// First check on "private"
+		pkg, err := registry.GetPackage(ctx, "private", parts[0], parts[1], version)
+		if err == nil {
+			return pkg, nil
+		} else if !errors.Is(err, ErrNotFound) && !errors.Is(err, ErrUnauthorized) && !errors.Is(err, ErrForbidden) {
+			return apitype.PackageMetadata{}, fmt.Errorf("unable to check on private/%s: %w", name, err)
+		}
+
+		// Then check on "pulumi"
+		pkg, err = registry.GetPackage(ctx, "pulumi", parts[0], parts[1], version)
+		if err == nil {
+			return pkg, nil
+		} else if !errors.Is(err, ErrNotFound) {
+			return apitype.PackageMetadata{}, fmt.Errorf("unable to check on pulumi/%s: %w", name, err)
+		}
+
+		// Both "private" and "pulumi" didn't exist, so we have successfully resolved to NotFound.
+		return apitype.PackageMetadata{}, fmt.Errorf("could not resolve %s: %w", name, ErrNotFound)
+	case 1:
+		var privatePackageMetadata, pulumiPackageMetadata *apitype.PackageMetadata
+		var suggested []apitype.PackageMetadata
+		for meta, err := range registry.SearchByName(ctx, &name) {
+			if err != nil {
+				return apitype.PackageMetadata{}, err
+			}
+
+			if meta.Source == "private" && privatePackageMetadata == nil {
+				privatePackageMetadata = &meta
+				break
+			} else if meta.Source == "pulumi" && meta.Publisher == "pulumi" {
+				pulumiPackageMetadata = &meta
+				// We don't break here, since we could still match a dominant source: the key in "private".
+			} else {
+				suggested = append(suggested, meta)
+			}
+		}
+
+		applyVersion := func(meta apitype.PackageMetadata) (apitype.PackageMetadata, error) {
```

Comment:
Why is this necessary?  Wouldn't `SearchByName` already return all packages with their versions? Seems a bit odd to me that we need to call `GetPackage` again here.

Reply:
`SearchByName` returns metadata on all packages *at their latest version*. If the version specified isn't the latest version, we need to call `GetPackageName` to retrieve the package at the version specified.